Concurrent tablet migration and balancing

tgrabiec opened the pull request:
This change makes tablet load balancing more efficient by performing
migrations independently for different tablets, and making new load
balancing plans concurrently with active migrations. 

The migration track is interrupted by pending topology change operations.

The coordinator executes the load balancer on edges of tablet state
machine transitions. This allows new migrations to be started as soon
as tablets finish streaming.

The load balancer is also continuously invoked as long as it produces
a non-empty plan. This is in order to saturate the cluster with
streaming. A single make_plan() call is still not saturating, due
to the way algorithm is implemented.

Overload of shards is limited by the fact that load balancer algorithm tracks
streaming concurrency on both source and target shards of active
migrations and takes concurrency limit into account when producing new
migrations.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2709/

avikivity commented on service/storage_service.cc:
```diff
@@ -1817,7 +1822,7 @@ future<> topology_coordinator::run() {
                 slogger.trace("raft topology: topology coordinator fiber has nothing to do. Sleeping.");
-                co_await _topo_sm.event.when();
+                co_await await_event();
                 slogger.trace("raft topology: topology coordinator fiber got an event");
```
I suggest to restructure the loop later in the conventional condition variable style.

avikivity commented on service/storage_service.cc:
```diff
@@ -1181,116 +1171,169 @@ class topology_coordinator {
+
+    // Set to true when any action started on behalf of a background_action_holder
+    // for any tablet finishes, or fails and needs to be restarted.
+    bool _tablets_ready = false;
```
Should it be a set of tablet identifiers? So we can quickly address the event?

tgrabiec replied:
Not yet.

avikivity commented on service/storage_service.cc:
```diff
@@ -1181,116 +1171,169 @@ class topology_coordinator {
+            holder = futurize_invoke(action)
+                        .finally([this, g = _async_gate.hold(), gid, name] () noexcept {
+                slogger.trace("raft topology: {} for tablet {} resolved.", name, gid);
+                _tablets_ready = true;
```
e.g. _tablets_ready.insert(gid).

If you don't want to handle failures here, you can pre-allocate the node to insert and just merge it here, using std::unordered_set::merge.

Yet another option is to use bitsets with summaries.

Of course, as an optimization, it can be deferred.

tgrabiec replied:
I actually started with a set, but it turned out we I don't need it because scheduling is done in the load balancer, which is stateless. If it was stateful, then it would make sense to notify it about changed. 

The nice thing about load balancer being stateless is that the code is very simple and easy to reason about.

avikivity replied:
It won't work when we manage 100k tablets.

avikivity commented on service/storage_service.cc:
```diff
@@ -1181,116 +1171,169 @@ class topology_coordinator {
+        // This step acts like a pump which advances state machines of individual tablets,
+        // batching barriers and group0 updates.
+        // If progress cannot be made, e.g. because all transitions are streaming, we block
+        // and wait for notification.
```
In a way it's similar to the reactor, waiting for futures to complete and launching the next tasks, but here we inject a big wrapper around each task.

avikivity commented on service/storage_service.cc:
```diff
@@ -1181,116 +1171,169 @@ class topology_coordinator {
+                    transition_to_with_barrier(locator::tablet_transition_stage::use_new);
+                    break;
+                case locator::tablet_transition_stage::use_new:
+                    transition_to_with_barrier(locator::tablet_transition_stage::cleanup);
```
We'd schedule a cleanup operation here, no? Maybe leave a comment?

tgrabiec replied:
Yes. I'll leave a fixme.

avikivity commented on docs/dev/topology-over-raft.md:
```diff
@@ -78,20 +78,20 @@ check if we need to rebalance. If so, it computes an incremental tablet migratio
-and thus limit the speed of load balancing.
+The tablet migration track excludes with other topology changes, so node operations
+will have to wait for tablet migration track to finish before they can take over
+the state machine.
```
Will node operations be prioritized? Where is this made explicit in the code?

tgrabiec replied:
They are. Tablet load balancing is stopped when we notice pending node operation:

```

        // In order to keep the cluster saturated, ask the load balancer for more transitions.
        // Unless there is a pending topology change operation.
        auto [preempt, new_guard] = should_preempt_balancing(std::move(guard));
        guard = std::move(new_guard);
        if (!preempt) {
            auto plan = co_await balance_tablets(get_token_metadata_ptr());
            co_await generate_migration_updates(updates, guard, plan);
        }
```

avikivity commented on service/storage_service.cc:
```diff
@@ -1333,6 +1333,15 @@ class topology_coordinator {
         }
 
+        // In order to keep the cluster saturated, ask the load balancer for more transitions.
+        // Unless there is a pending topology change operation.
```
Ok it's here.

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -120,7 +133,65 @@ class load_balancer {
+    // We allow at least two sessions per shard so that there is less chance for idling until load balancer
+    // makes the next decision after streaming is finished.
+    const size_t max_write_streaming_load = 2;
+    const size_t max_read_streaming_load = 4;
```
I feel that 2 would be better here, but it's not based on anything concrete.

Later, we can vary it by the actual load on the node. A newly-bootstrapped node hasn't much work to do so we can allow it to stream more tablets from the entire cluster. As its load increases, we reduce the stream-in concurrency to make more room for work on the tablets it gained.

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -212,42 +288,46 @@ class load_balancer {
         // If target node is not balanced across shards, we will overload some shards.
         // If tablets are not balanced in size, throughput will suffer because some shards will be idle sooner than others.
```
Outdated?

tgrabiec replied:
Right.

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -212,42 +288,46 @@ class load_balancer {
-        // by the target shard being busy doesn't mean we could not proceed with migrations for other shards
-        // which would be produced by the planner subsequently.
+        // FIXME: To handle the above, we should rebalance the target node before migrating tablets from other nodes.
+        // If shards are balanced on the target node, the balancer will naturally distribute tablets to different shards.
```
If there is mild imbalance, won't it be corrected by the algorithm? It will choose less-loaded shards first.

We should reject a shard as a target if its load >= average load, as migrating a tablet to it will just cause it to get migrated back later.

tgrabiec replied:
> If there is mild imbalance, won't it be corrected by the algorithm? It will choose less-loaded shards first.

Yes, it will, we fill least-loaded shards first. 

The problem is that imbalance will mean that we will not choose more loaded shards, so streaming capacity is reduced as those shards don't participate. If we equalized intra-node load first, which should be cheap, all shards could participate.

If the imbalance is mild and is corrected quickly, it's not an issue.

avikivity replied:
Sure, but I see it as a second-order problem. Internal imbalance will be due to resharding, which is rare.

btw, we should aim to have good distribution of tables. Avoid table A on shard 0 and table B on shard 1, and instead have AB balanced on 01, in proportion to their sizes (and a similar policy for nodes).

tgrabiec replied:
Yes, it is a second-order problem. It will get more serious when we base load on dynamic metrics like hotness.

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -385,8 +467,30 @@ class load_balancer {
+                    lblogger.debug("Too many migrations skipped, aborting balancing");
+                    break;
+                }
+            }
```
We should be careful with exceptions here. If we get an exception, then recovery should be sufficiently outer to this code so all the data structures are re-created, since we won't have strong exception guarantees with so many structures.

tgrabiec replied:
Currently, there are no exception guarantees here, so we're good.

avikivity replied:
But we where is the recovery code? I guess it's shared with overall topology coordination.

tgrabiec replied:
The algorithm itself is stateless, so there is no need for recovery here. There are no side-effects if the plan is not executed. 

On exception, the topology coordinator will retry starting from the current group0 state.

avikivity commented:
Very good.

tgrabiec commented:
In V2:
  - Added a commit which removes duplicated log line
  - Fixed outdated comment in make_plane() which refers to reduced throughput due to streaming tablets in tandem:
    ```patch
    @@ -209,45 +285,50 @@ class load_balancer {
             // We want to saturate the target node so we migrate several tablets in parallel, one for each shard
             // on the target node. This assumes that the target node is well-balanced and that tablet migrations
             // complete at the same time. Both assumptions are not generally true in practice, which we currently ignore.
    -        // If target node is not balanced across shards, we will overload some shards.
    -        // If tablets are not balanced in size, throughput will suffer because some shards will be idle sooner than others.
    -        //
    -        // FIXME: To handle the above, we should (1) rebalance the target node
    -        // before migrating tablets from other nodes. If shards are balanced on the target node, the balancer
    -        // will naturally distribute tablets to different shards. Also, (2) we should change this algorithm
    -        // to be a generator for migrations and have a scheduler in the execution layer which pulls migrations
    -        // from this algorithm, batches them and decides how many to execute.
    +        // But they will be true typically, because we fill shards starting from least-loaded shards,
    +        // so we naturally strive towards balance between shards.
             //
    -        // The scheduler decides in which order to execute the plan based on current activity in the system.
    -        // We cannot just ask the planner for the next migration and stop when we hit overload on some shard,
    -        // because that can lead to underutilization of the cluster. Just because the next migration is blocked
    -        // by the target shard being busy doesn't mean we could not proceed with migrations for other shards
    -        // which would be produced by the planner subsequently.
    +        // If target node is not balanced across shards, we will overload some shards. Streaming concurrency
    +        // will suffer because more loaded shards will not participate, which will under-utilize the node.
    +        // FIXME: To handle the above, we should rebalance the target node before migrating tablets from other nodes.
    ```
  - Preserve FIXME about lack of cleanup

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2758/

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1744,6 +1744,11 @@ class topology_coordinator {
     future<bool> maybe_start_tablet_migration(group0_guard);
+
+    future<> await_event() {
+        _as.check();
```
But we have this at the beginning of the coordinator:

```
    auto abort = _as.subscribe([this] () noexcept {
        _topo_sm.event.signal();
    });
```

So in the scenario you describe (if I understand it correctly) the coordinator should be woken.

tgrabiec replied:
It's not enough, because this signal() can be executed before the coordinator executes await().

tgrabiec replied:
Right. Although, this implementation details is not documented. Should be document it and rely on it? @avikivity

avikivity replied:
It's not good to rely on this (condition_variable remembering being woken). It's not how standard condition variables work.

avikivity replied:
In general the loop to wait for the condition variable should check any state for which can cause a wakeup.

tgrabiec replied:
But then we have to emulate this behavior outside of condition variable, like I did with _tablets_ready, or by taking a lock (group0 guard) and atomically releasing it with waiting.

avikivity replied:
If you'd need to do that with a standard condition variable, then that's okay. If we use the name of a standard thing, then it should have the semantics of the standard thing.

avikivity replied:
There is no real condition variable, only documentation.
